### PR TITLE
feat(provider): add bodyTemplate for POST body-based pagination

### LINKED ISSUE #330: feat(http): Support POST body-based pagination (GraphQL, Elasticsearch)
## Problem

The http provider's pagination strategies (`pageNumber`, `offset`, `cursor`, `linkHeader`, `custom`) all operate on URL query parameters or response link headers. None of them can rewrite the **request body** between pages.

This means POST-based paginated APIs cannot use built-in pagination. Users must work around it with 3+ extra resolvers and a `forEach` pattern (~30 lines of YAML for what should be a pagination config option).

## Affected Use Cases

- **GraphQL APIs** (Apollo, Hasura, AWS AppSync, GitHub GraphQL) -- pagination params are always in the POST body
- **Elasticsearch/OpenSearch** `_search` endpoint -- uses POST with `from`/`size` or `search_after` in the body
- **Any enterprise REST API** with POST-based search/query endpoints

This is arguably the most common pagination pattern in enterprise environments.

## Proposed Solution

Add a `bodyTemplate` (or `nextBody`) option to the pagination config that supports page context variables:

```yaml
pagination:
  strategy: pageNumber
  maxPages: 89
  pageSize: 100
  bodyTemplate:
    expr: |
      json.marshal({
        "query": "{ items(paginationOptions: {size: " + string(__pageSize) + ", page: " + string(__page) + "}) { results { id name } totalItems } }"
      })
  collectPath: "body.data.items.results"
  stopWhen: "size(body.data.items.results) == 0"
```

### Context Variables

| Variable | Description |
|----------|-------------|
| `__page` | Current page number (1-indexed) |
| `__pageSize` | Configured page size |
| `__cursor` | Cursor/token from previous response (cursor strategy) |
| `__offset` | Current offset (offset strategy) |

When `bodyTemplate` is set, the pagination engine regenerates the request body on each iteration instead of modifying URL query params.

## Design Considerations

- **Interaction with `body` input**: When `bodyTemplate` is set, it should override the top-level `body` on pages 2+. Page 1 could use either the original `body` or the template.
- **Variable namespace**: Use `__` prefix (consistent with `__self`, `__plan`, `__actions`).
- **Strategy compatibility**: Should work with all strategies -- `pageNumber` (inject `__page`), `offset` (inject `__offset`), `cursor` (inject `__cursor`).

## Current Workaround

Requires 3+ extra resolvers with `forEach`:

1. Fetch page 1 to get `totalPages`
2. Generate page numbers array: `lists.range(totalPages).map(i, i + 1)`
3. `forEach` on transform step to fetch each page with concurrency
4. Flatten results with another CEL transform

## Context

Discovered while building a solution querying a GraphQL API that uses standard page-number pagination in the POST body. Every GraphQL API works this way.

### PR DESCRIPTION
**Pull request by kcloutie**
- add bodyTemplate field to pagination config, evaluated per-page via CEL to generate dynamic request bodies for POST-based APIs
- expose __page, __pageSize, __offset, __cursor context variables for use within bodyTemplate expressions
- validate that bodyTemplate requires POST, PUT, or PATCH method
- support string results (sent as-is) and map/list results (auto- marshaled to JSON)
- add pageSize() helper to paginationConfig for strategy-aware size
- guard offset/cursor state tracking to only run when bodyTemplate is configured
- add unit tests for pageNumber, offset, and cursor strategies with bodyTemplate, plus validation error and nil/map result branches
- add integration test with mock HTTP service verifying multi-page collection
- add example solution with GraphQL, Elasticsearch, cursor, and map body patterns
- document body-based pagination in docs/design/providers.md

Closes #330

### CODE REVIEW

File: pkg/provider/builtin/httpprovider/pagination.go [hunk 1]
```diff
@@ -297,6 +339,20 @@ func (p *HTTPProvider) executePaginated(
...
+					if s, ok := cursor.(string); ok {
+						lastCursor = s
+					} else {
+						lastCursor = fmt.Sprintf("%v", cursor)
+					}
+				}
+			}
+		}
+
 		currentURL = nextURL
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply pkg/provider/builtin/httpprovider/pagination.go]: Fixed: xecutePaginated now skips 
esolveNextPage entirely when odyTemplate is set. The URL remains unchanged across all pages. Added URL-stability assertions to all bodyTemplate tests.

File: pkg/provider/builtin/httpprovider/pagination.go [hunk 2]
```diff
@@ -218,10 +247,23 @@ func (p *HTTPProvider) executePaginated(
...
+	var lastCursor string
+	currentOffset := 0
+
 	for pageCount < pagCfg.MaxPages {
 		pageCount++
 		lgr.V(1).Info("fetching page", "provider", ProviderName, "page", pageCount, "url", currentURL)
 
+		// Evaluate bodyTemplate if configured, replacing the request body for this page.
+		if pagCfg.BodyTemplate != "" {
+			newBody, err := p.evaluateBodyTemplate(ctx, pagCfg, pageCount, currentOffset, lastCursor)
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply pkg/provider/builtin/httpprovider/pagination.go]: Fixed: __page now starts at startPage (default 1) and __offset is seeded from the initial URL's offset query param via new initialOffset() helper. Added tests covering both behaviors.

File: pkg/provider/builtin/httpprovider/pagination.go [hunk 3]
```diff
@@ -73,6 +86,19 @@ func defaultPaginationConfig() paginationConfig {
...
 }
 
+// pageSize returns the effective page size for the configured strategy.
+// Returns Limit for offset, PageSize for pageNumber, 0 for strategies without a fixed page size.
+func (c *paginationConfig) pageSize() int {
+	switch c.Strategy {
+	case StrategyOffset:
+		return c.Limit
+	case StrategyPageNumber:
+		return c.PageSize
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply pkg/provider/builtin/httpprovider/pagination.go]: Fixed: pageSize is now parsed for the cursor strategy too, and pageSize() returns it as a fallback for all non-offset/non-pageNumber strategies. This makes __pageSize available in cursor+bodyTemplate mode. Added TestParsePaginationConfig_CursorWithPageSize test.

File: pkg/provider/builtin/httpprovider/http.go [hunk 4]
```diff
@@ -229,6 +229,9 @@ func NewHTTPProvider() *HTTPProvider {
 					"collectPath": schemahelper.StringProp("CEL expression to extract items from each page's response body. Items are accumulated across pages into a single array.",
 						schemahelper.WithExample("body.items"),
 						schemahelper.WithMaxLength(*ptrs.IntPtr(500))),
+					"bodyTemplate": schemahelper.StringProp("CEL expression evaluated per-page to generate the request body. Enables POST-based pagination (GraphQL, Elasticsearch). When set, overrides the top-level body field. Available variables: __page (1-indexed), __pageSize, __offset, __cursor.",
+						schemahelper.WithExample(`'{"query":"{ items(page: ' + string(__page) + ', size: ' + string(__pageSize) + ') { results { id } } }"}'`),
+						schemahelper.WithMaxLength(*ptrs.IntPtr(5000))),
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply pkg/provider/builtin/httpprovider/http.go]: Fixed: Increased odyTemplate maxLength from 5000 to 50000 to accommodate complex GraphQL queries with variables, fragments, etc.

File: pkg/provider/builtin/httpprovider/pagination_test.go [hunk 5]
```diff
@@ -1260,3 +1261,449 @@ func TestHTTPProvider_Pagination_CollectPath_MissingKeyOnLastPage(t *testing.T)
...
+		"url":    server.URL + "/graphql",
+		"method": "POST",
+		"pagination": map[string]any{
+			"strategy":     "pageNumber",
+			"maxPages":     10,
+			"pageSize":     pageSize,
+			"bodyTemplate": `'{"page": ' + string(__page) + ', "size": ' + string(__pageSize) + '}'`,
+			"collectPath":  "body.data.items",
+			"stopWhen":     "size(body.data.items) == 0",
+		},
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply pkg/provider/builtin/httpprovider/pagination_test.go:1387]: Fixed: All three bodyTemplate tests (PageNumber, Offset, Cursor) now capture received URLs and assert they remain unchanged. Cursor test also verifies body content carries correct cursor values across pages.

File: tests/integration/solutions/providers/http-body-pagination/solution.yaml [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      services:
+        - name: mock-api
+          type: http
+          portEnv: MOCK_HTTP_PORT
+          baseUrlEnv: MOCK_BASE_URL
+          routes:
+            - path: /api/items
+              method: POST
+              status: 200
+              body: '{"items":[{"id":1,"name":"alpha"},{"id":2,"name":"beta"}]}'
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply tests/integration/solutions/providers/http-body-pagination/solution.yaml:61]: Fixed: Integration test now uses odyContains routing to verify each page sends a different body. Three separate routes match page 1, 2, and 3 bodies, returning distinct response data.

File: docs/design/providers.md [hunk 7]
```diff
@@ -1074,6 +1074,69 @@ resolve:
         method: GET
 ~~~
 
+#### Body-Based Pagination
+
+For POST-based paginated APIs (GraphQL, Elasticsearch, enterprise search), use
+`bodyTemplate` in the pagination config. This CEL expression is evaluated
+per-page to generate a new request body for each iteration.
+
+Available context variables in `bodyTemplate`:
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply docs/design/providers.md:1083]: Fixed: Added a full 'Body-Based Pagination (bodyTemplate)' section to docs/tutorials/provider-reference.md with field table, context variable documentation, and GraphQL/Elasticsearch examples.

File: pkg/provider/builtin/httpprovider/pagination.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kcloutie) [reply pkg/provider/builtin/httpprovider/pagination.go]: Addressed: URL no longer mutates when bodyTemplate is set. The example has been updated to show pageSize: 50 with __pageSize in the expression.

File: examples/providers/http-body-pagination.yaml [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+              headers:
+                Content-Type: application/json
+                Authorization: "Bearer ${GRAPHQL_TOKEN}"
+              pagination:
+                strategy: pageNumber
+                maxPages: 50
+                pageSize: 100
+                bodyTemplate: |
+                  '{"query":"{ items(page: ' + string(__page) + ', size: ' + string(__pageSize) + ') { results { id name createdAt } totalPages } }"}'
+                collectPath: "body.data.items.results"
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply examples/providers/http-body-pagination.yaml:43]: Fixed: URL remains unchanged in bodyTemplate mode. The 
esolveNextPage logic is now only invoked for URL-based pagination strategies.

File: docs/design/providers.md [hunk 9]
```diff
@@ -1074,6 +1074,69 @@ resolve:
...
+
+| Variable | Description |
+|----------|-------------|
+| `__page` | Current page number (1-indexed) |
+| `__pageSize` | Configured page size (from `pageSize` or `limit`) |
+| `__offset` | Current offset (increments by page size each page) |
+| `__cursor` | Cursor token from previous response (cursor strategy) |
+
+When `bodyTemplate` is set, it overrides the top-level `body` field for all
+pages.
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply docs/design/providers.md]: Addressed in the same fix as the URL mutation issue. 
esolveNextPage is skipped entirely when odyTemplate is set.

File: pkg/provider/builtin/httpprovider/http.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (kcloutie) [reply pkg/provider/builtin/httpprovider/http.go]: Fixed: Added odyTemplate assertion to TestHTTPProvider_Pagination_DescriptorHasPaginationSchema. Also updated docs to clarify URL immutability.

File: pkg/provider/builtin/httpprovider/pagination.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kcloutie) [reply pkg/provider/builtin/httpprovider/pagination.go]: Addressed with Fix #2: __page now starts at startPage and __offset seeds from the initial URL. Tests added for initialOffset() covering all edge cases.

File: pkg/provider/builtin/httpprovider/pagination.go [hunk 10]
```diff
@@ -218,10 +247,23 @@ func (p *HTTPProvider) executePaginated(
 	pageCount := 0
 	currentURL := urlStr
 
+	// Track state for bodyTemplate context variables
+	var lastCursor string
+	currentOffset := 0
+
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply pkg/provider/builtin/httpprovider/pagination.go]: Addressed with the __page/__offset initialization fix. The currentPage variable starts from pagCfg.StartPage and currentOffset from initialOffset().

File: pkg/provider/builtin/httpprovider/pagination_test.go [hunk 11]
```diff
@@ -1260,3 +1261,449 @@ func TestHTTPProvider_Pagination_CollectPath_MissingKeyOnLastPage(t *testing.T)
...
+		"url":    server.URL + "/graphql",
+		"method": "POST",
+		"pagination": map[string]any{
+			"strategy":       "cursor",
+			"maxPages":       10,
+			"nextTokenPath":  "body.nextCursor",
+			"nextTokenParam": "cursor",
+			"bodyTemplate":   `'{"after": "' + __cursor + '"}'`,
+			"collectPath":    "body.items",
+		},
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply pkg/provider/builtin/httpprovider/pagination_test.go:1554]: Fixed: Cursor test now asserts body content (cursor values per page) and URL immutability.

File: tests/integration/solutions/providers/http-body-pagination/solution.yaml [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+              headers:
+                Content-Type: application/json
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply tests/integration/solutions/providers/http-body-pagination/solution.yaml:63]: Addressed: Integration test now uses odyContains routing with distinct per-page responses, verifying that different bodies are sent for each page.

File: docs/design/providers.md [hunk 13]
```diff
@@ -1074,6 +1074,69 @@ resolve:
...
+
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply docs/design/providers.md:1095]: Done: docs/tutorials/provider-reference.md now has a full bodyTemplate section with field table, variable docs, and examples.

File: pkg/provider/builtin/httpprovider/pagination.go [hunk 14]
```diff
@@ -287,17 +359,38 @@ func (p *HTTPProvider) executePaginated(
...
+					if s, ok := cursor.(string); ok {
+						lastCursor = s
+					} else {
+						lastCursor = fmt.Sprintf("%v", cursor)
+					}
+				} else {
+					// No cursor returned — no more pages.
+					lgr.V(1).Info("pagination completed", "page", pageCount, "reason", "cursor exhausted")
+					break
+				}
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply pkg/provider/builtin/httpprovider/pagination.go]: Fixed: The bodyTemplate cursor branch now uses isNullOrEmpty() (which handles structpb.NullValue) to properly detect CEL null as end-of-pagination, matching the behavior of 
esolveCursorNext.

[the remark this answers is not in the document]
Comment (kcloutie) [reply pkg/provider/builtin/httpprovider/pagination.go]: Fixed: Evaluation errors from 
extTokenPath are now surfaced as proper errors (mt.Errorf with context) instead of being silently treated as a stop condition. This matches the error-reporting behavior of 
esolveCursorNext.

File: pkg/provider/builtin/httpprovider/pagination.go [hunk 15]
```diff
@@ -663,6 +756,42 @@ func evaluateCELForPagination(ctx context.Context, expression string, responseCt
...
+	switch v := result.(type) {
+	case string:
+		return v, nil
+	case nil:
+		return "", nil
+	default:
+		// Non-string results (maps, slices) are marshaled to JSON
+		jsonBytes, err := json.Marshal(v)
+		if err != nil {
+			return "", fmt.Errorf("failed to marshal bodyTemplate result to JSON: %w", err)
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply pkg/provider/builtin/httpprovider/pagination.go:924]: Fixed: valuateBodyTemplate now uses isNullOrEmpty(result) which properly detects structpb.NullValue (CEL null). When the template evaluates to null, an empty body is sent. The test now asserts odySize == 0 for each request.

File: pkg/provider/builtin/httpprovider/pagination.go [hunk 16]
```diff
@@ -168,6 +217,13 @@ func parsePaginationConfig(inputs map[string]any) (*paginationConfig, error) {
 		if nup, ok := pagMap["nextURLPath"].(string); ok {
 			cfg.NextURLPath = nup
 		}
+		// Allow pageSize for cursor strategy (used by bodyTemplate's __pageSize)
+		if ps, ok := pagMap["pageSize"].(int); ok && ps > 0 {
+			cfg.PageSize = ps
+		}
+		if ps, ok := pagMap["pageSize"].(float64); ok && ps > 0 {
+			cfg.PageSize = int(ps)
+		}
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply pkg/provider/builtin/httpprovider/pagination.go:226]: Fixed: pageSize is now parsed in the linkHeader and custom strategy branches as well, so __pageSize is available in bodyTemplate mode for all strategies.

File: pkg/provider/builtin/httpprovider/pagination_test.go [hunk 17]
```diff
@@ -1260,3 +1262,528 @@ func TestHTTPProvider_Pagination_CollectPath_MissingKeyOnLastPage(t *testing.T)
...
+	}
+
+	output, err := p.Execute(ctx, inputs)
+	require.NoError(t, err)
+	require.NotNil(t, output)
+
+	// Verify it ran (the nil body should have been sent as empty string)
+	data := output.Data.(map[string]any)
+	assert.NotNil(t, data["body"])
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply pkg/provider/builtin/httpprovider/pagination_test.go]: Fixed: Test now includes odySize in each collected item and asserts odySize == 0 for every page, ensuring dyn(null) actually sends an empty body rather than a marshaled payload.

File: examples/providers/http-body-pagination.yaml [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+              method: POST
+              headers:
+                Content-Type: application/json
+                Authorization: "Bearer ${GITHUB_TOKEN}"
+              pagination:
+                strategy: cursor
+                maxPages: 20
+                pageSize: 50
+                nextTokenPath: "body.data.viewer.repositories.pageInfo.endCursor"
+                nextTokenParam: "after"
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply examples/providers/http-body-pagination.yaml]: Fixed: 
extTokenParam is no longer required when odyTemplate is set — the validation now only enforces it for URL-based cursor pagination. Removed the dummy 
extTokenParam from the example and docs.

File: docs/tutorials/provider-reference.md [hunk 19]
```diff
@@ -1529,6 +1529,61 @@ Full control using CEL expressions.
...
+| Field | Type | Description |
+|-------|------|-------------|
+| `bodyTemplate` | string | CEL expression evaluated per-page to produce the request body. Overrides the top-level `body` field. |
+
+**CEL variables available** in `bodyTemplate`:
+
+| Variable | Type | Description |
+|----------|------|-------------|
+| `__page` | int | Current page number (starts at `startPage`, default 1) |
+| `__pageSize` | int | Page size from `pageSize` (pageNumber) or `limit` (offset) |
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply docs/tutorials/provider-reference.md]: Fixed: Updated __pageSize description to state it comes from pageSize for pageNumber/cursor/linkHeader/custom strategies (or limit for offset), matching the actual behavior now that all strategy branches parse it.

File: pkg/provider/builtin/httpprovider/pagination.go [hunk 20]
```diff
@@ -287,17 +373,40 @@ func (p *HTTPProvider) executePaginated(
...
-		nextURL, stop, err := p.resolveNextPage(ctx, pagCfg, currentURL, resp, responseCtx)
-		if err != nil {
-			return nil, fmt.Errorf("%s: page %d failed to resolve next page: %w", ProviderName, pageCount, err)
-		}
-		if stop {
-			lgr.V(1).Info("pagination completed", "page", pageCount, "reason", "no more pages")
-			break
+		// When bodyTemplate is set, pagination state lives in the body — do not
+		// mutate the URL. Otherwise, use the strategy to resolve the next URL.
+		if pagCfg.BodyTemplate != "" {
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply pkg/provider/builtin/httpprovider/pagination.go:380]: Fixed: The bodyTemplate branch now evaluates response-driven stop signals per strategy using new `hasLinkNextHeader()` and `checkCustomStop()` helpers (no host validation since the URL is never followed). linkHeader checks for Link rel=next, custom evaluates nextURL/nextParams, cursor checks nextURLPath/nextTokenPath. Added tests covering each strategy stop signal in bodyTemplate mode.

File: pkg/provider/builtin/httpprovider/pagination.go [hunk 21]
```diff
@@ -287,17 +373,40 @@ func (p *HTTPProvider) executePaginated(
...
+			// Update state for bodyTemplate context variables on the next iteration.
+			currentPage++
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply pkg/provider/builtin/httpprovider/pagination.go]: Fixed: The bodyTemplate branch now applies `checkItemCountStop()` for offset/pageNumber strategies (on page 2+), so bare-array responses with fewer items than pageSize correctly terminate pagination without needing an explicit `stopWhen`. Added `TestHTTPProvider_Pagination_BodyTemplate_ShortPage_StopSignal` to cover this path.

File: docs/tutorials/provider-reference.md [hunk 22]
```diff
@@ -1529,6 +1529,60 @@ Full control using CEL expressions.
 | `nextURL` | string | CEL expression returning the full next page URL (empty string = stop) |
 | `nextParams` | string | CEL expression returning a map of query params for the next request (empty map = stop) |
 
+#### Body-Based Pagination (bodyTemplate)
+
+For APIs that paginate via the request body (GraphQL, Elasticsearch), use `bodyTemplate` to generate the POST body for each page instead of modifying URL query parameters.
+
+| Field | Type | Description |
+|-------|------|-------------|
+| `bodyTemplate` | string | CEL expression evaluated per-page to produce the request body. Overrides the top-level `body` field. |
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply docs/tutorials/provider-reference.md]: Fixed: Updated the bodyTemplate field description in provider-reference.md to explicitly state that the expression may return a string (sent as-is) or a map/list (automatically JSON-marshaled), and that null produces an empty body.

File: docs/tutorials/provider-reference.md [hunk 23]
```diff
@@ -1529,6 +1529,60 @@ Full control using CEL expressions.
...
+  method: POST
+  headers:
+    Content-Type: application/json
+  pagination:
+    strategy: cursor
+    maxPages: 10
+    pageSize: 50
+    nextTokenPath: "body.data.pageInfo.endCursor"
+    bodyTemplate: |
+      '{"query":"{ items(first: ' + string(__pageSize) + ', after: \"' + __cursor + '\") { nodes { id } pageInfo { endCursor hasNextPage } } }"}'
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply docs/tutorials/provider-reference.md]: Fixed: The GraphQL cursor example in provider-reference.md now uses a conditional expression (`__cursor == ""` ternary) to omit the cursor argument on the initial request instead of sending an empty string, matching the pattern already used in examples/providers/http-body-pagination.yaml.

File: docs/tutorials/provider-reference.md [hunk 24]
```diff
@@ -1529,6 +1529,60 @@ Full control using CEL expressions.
...
+  url: https://api.example.com/graphql
+  method: POST
+  headers:
+    Content-Type: application/json
+  pagination:
+    strategy: cursor
+    maxPages: 10
+    pageSize: 50
+    nextTokenPath: "body.data.pageInfo.endCursor"
+    bodyTemplate: |
```
[the remark this answers is not in the document]
Comment (kcloutie) [reply docs/tutorials/provider-reference.md:1564]: Fixed: Updated the cursor field table to note that `nextTokenParam` is required with `nextTokenPath` *unless* `bodyTemplate` is set, clarifying the exception for body-based pagination.
